rfcs/0012-pipeline-executor.md

```diff
@@ -0,0 +1,367 @@
+    mconfig3["target"] = "llvm"
+    mconfig3["dev"] = tvm.cpu(0)
+
+    mconfig3["pipeline"] = {
+        "mod_indx": 3,
+        "output": [{"output_indx": 1, "dependent": [{"mod_indx": 0, "input_name": "2"}]}],
+    }
+    mod_config[mods[2]] = mconfig3
+``` 
+
+### 3.3. Use pipeline_executor to build pipeline module with the said subgraph and configuration.
+
+Pipeline executor provide a build function to compile and save the compile output into disk,
+following is a example
+
+```python
+    with relay.build_config(opt_level=3):
+        pipeline_mods, string_config = pipeline_executor.build_pipeline(
+            mod_config, "<path to storage the build output>"
+        )
+
+```
+
+### 3.4. Use pipeline_executor to load pipeline module to run network in pipeline parallism mode.
+
+Pipeline executor working at asynchronous mode, unlike the synchronous mode set_input->run->get_out
+sequence, Pipeline executor execte in such order for example.
+    set_input--> run
+    set_input--> run
+    get_ouput
+    set_input-->run
+    get_output
+    get_output
+
+get_output can be called before set_input/run, and it may return a empy array that means no output ready
+or a valid output.
+
+following is one example
+
+```python
+#...
+
+def get_output(outputs, module):
+  suc = False
+  output = pipeline_module.get_output()
+  if len(output):
+    curOutputs = [output.asnumpy() for data in output]
+    outputs.append(curOutputs)
+    suc = True
+
+  return suc
+    
+
+pipeline_outputs = []
+datas = []
+for i in range(len(mods) + 1):
+  datas.append(np.full(dshape, 3 + i).astype("float32"))
+pipeline_module = pipeline_executor.create(pipeline_mods, string_config)
+
+for data in datas:
+    get_output(pipeline_outputs, pipeline_module)
+    pipeline_module.set_input("data_0", data)
+    pipeline_module.set_input("data_1", data, mod_idx=2)
+    pipeline_module.run()
+    get_output(pipeline_outputs, pipeline_module)
+
+left = len(datas) - len(pipeline_outputs)
+while(left > 0):
+    left = left - 1 if get_output(pipeline_outputs, pipeline_module) else left
+```
+
+## 4 Reference-level explanation
+This section introduces the underlying techniques for the pipeline executor.
+The figure below briefly illustrates the workflow of the system
+
+Pipeline executor architecture
+![meta-schedule-workflow](../resources/pipeline-executor-arch.png)
+
+Manually construct the subgraph
+![meta-schedule-workflow](../resources/pipeline-executor-subgraph-split.png)
+
+How pipeline executor runtime work
+![meta-schedule-workflow](../resources/pipeline-executor-runtime.png)
+
+The pipeline executor schedule logic
+![meta-schedule-workflow](../resources/pipeline-executor-schedule.png)
+
+The network pipeline compute effect
+![meta-schedule-workflow](../resources/pipeline-executor-pipeline.png)
+
+
+## 5. Drawbacks
+
+
+Pipeline executor currently need manually subgraph split and configuration construct,
+Further graph split feature would doing automatically split.
```

Comment:
```suggestion
Pipeline executor currently needs manually subgraph splitting and configuration construction.
Further graph splitting feature would do automatically split.
```

I would say this is an important feature for this RFC, so it might be better to include possible solutions even it is not going to be implemented for now. Without this feature, I don't really believe pipeline executor itself would be useful for the community.

Reply:
sure, about graph split part, following are the initial solution,  this solution include 3 steps, 1. Operator Auto tune,
2. Graph dependency tree build and balance, 3. Graph Auto Tune. following are more detail.

1. Operator Auto Tune : 
    1.1 In operator Auto tune  tune section, user would using existing tuning logic to tune the every operator,
          but the tune would separately and serialized happen in all target involved by pipeline executor.
    1.2 After operator tune done ,  we would get performance data, for example , con2d_0 best perf in 
         GPU is 3ms, in VTA is 2ms etc, this perf data would get used in later graph split logic.
2.  Graph dependency tree build balance 
     2.1 Initialize a DAG, the node of the DAG is subgraph, initially for a N node DAG, first [1, N -1] node mapping to
           [1 , N-1] layer(compute density operator and others) of original compute graph, the number N node is
           mapping to [N, M] layer , M here is the original compute layer number.
     2.2, by useing the perf data generated in 1.2 , every dependency tree node can get a time consume value,  
           we would do the balance work to adjust the node subgraph scope to make every node have close perf 
           value.
3.  Graph Auto Tune.
     2. can generate more than one subgraph split solution, in this step, Graph Auto Tune would try these multiple
     solution to get best configuration.

after 1. 2. 3. , here can get an automatic graph split configuration.